File: solver/manifold.py

from geometry.primitives import Body
from .constraints import Constraint, ContactConstraint
import numpy as np

def _contact_key(point: np.ndarray, normal: np.ndarray, decimals: int = 1) -> tuple[tuple, tuple]:
    
    p_key = tuple(np.round(point.astype(float), decimals))
    n_key = tuple(np.round(normal.astype(float), decimals))
    return p_key, n_key


class Manifold():
    """
    Persistent container for constraints and contacts between bodies (a pair of bodies).
    """

    def __init__(self, A: Body, B: Body):
        self.A = A
        self.B = B

        self.constraint_dict : dict[int, ContactConstraint] = {}    # dictionary of { (p_key, n_key), ContactConstraint }
        self.separation_frames: dict[int, int]= {}                  # dictionary of { (p_key, n_key), int }

    @property
    def constraints(self) -> list[Constraint]:
        return list(self.constraint_dict.values())
    
    def update_from_contacts(self, contacts: list, friction: float, max_sep_frames: int) -> None:
        
        seen_keys = set()

        for c in contacts:
            print("Contact feature id: ", c.feature_id)
            key = c.feature_id

            if key in self.constraint_dict:
                print(f"\t\t!!Old constraint used for {c.bodyA.body_id} and {c.bodyB.body_id}.!!\n")
                cons = self.constraint_dict[key]
                cons.contact = c
                self.separation_frames[key] = 0
            else:
                print(f"\t\tNew constraint made for {c.bodyA.body_id} and {c.bodyB.body_id}.")
                cons = ContactConstraint(contact=c, friction=friction)
                self.constraint_dict[key] = cons
                self.separation_frames[key] = 0
            seen_keys.add(key)

        print("Seen keys: ", seen_keys)
        print("Current keys: ", self.constraint_dict.keys())

        to_delete = []

        for key in self.constraint_dict.keys():
            if key not in seen_keys:
            #     self.separation_frames[key] += 1
            #     if self.separation_frames[key] > max_sep_frames:
                to_delete.append(key)
        
        for key in to_delete:
            self.constraint_dict.pop(key, None)
            self.separation_frames.pop(key, None)

    def is_empty(self) -> bool:
        return len(self.constraint_dict) == 0

File: solver/Julia/manifold.jl
# manifold.jl
module ManifoldHandling

using LinearAlgebra
using StaticArrays

# Imports from your other modules
# Assuming Body is in Collisions and ContactConstraint is in Constraints
using ..Collisions: Body, Contact, get_aabb
using ..Constraints: ContactConstraint

export Manifold, update_manifold!, is_empty

# --- Constants ---
const MAX_SEP_FRAMES = 3

# --- Helper Functions (Private) ---

"""
Generates a spatial key for a contact based on its local position and normal.
Matches Python: _key_from_contact / _contact_key
"""
function _contact_key(point::SVector{3,Float64}, normal::SVector{3,Float64}; p_dec::Int=3, n_dec::Int=2)
    # Quantize point and normal for hash stability
    p_key = round.(point, digits=p_dec)
    n_key = round.(normal, digits=n_dec)
    # We return a tuple that Julia can hash automatically
    return (p_key, n_key)
end

function _proximity_test(c::Contact, old_con::ContactConstraint; dist_eps=2e-3, cos_eps=0.985)
    # Check distance squared
    dist_sq = sum(abs2, c.point - old_con.point)
    if dist_sq > dist_eps^2
        return false
    end

    # Check normal alignment
    if dot(c.normal, old_con.normal) < cos_eps
        return false
    end

    return true
end

function _aabb_overlap(A::Body, B::Body)
    # Conservative test using AABBs
    aabbA = get_aabb(A)
    aabbB = get_aabb(B)

    # Check for separation on any axis
    if aabbA.max[1] < aabbB.min[1] || aabbB.max[1] < aabbA.min[1]
        return false
    end
    if aabbA.max[2] < aabbB.min[2] || aabbB.max[2] < aabbA.min[2]
        return false
    end
    if aabbA.max[3] < aabbB.min[3] || aabbB.max[3] < aabbA.min[3]
        return false
    end

    return true
end

# --- Manifold Struct ---

mutable struct Manifold
    bodyA::Body
    bodyB::Body

    # Stores constraints mapped by spatial key
    # Dict{ KeyTuple, ContactConstraint }
    constraint_dict::Dict{Any,ContactConstraint}

    # Stores age of separation for keys
    # Dict{ KeyTuple, Int }
    separation_frames::Dict{Any,Int}

    # The active list exposed to the solver
    constraints::Vector{ContactConstraint}

    function Manifold(A::Body, B::Body)
        new(A, B,
            Dict{Any,ContactConstraint}(),
            Dict{Any,Int}(),
            ContactConstraint[])
    end
end

# --- Main Logic ---

"""
Updates the manifold with new contacts, handling warm-starting and persistence.
Matches Python: Manifold.update_from_contacts
"""
function update_manifold!(manifold::Manifold, new_contacts::Vector{Contact}, friction::Float64; dist_eps=0.1, cos_eps=0.95)
    old_dict = manifold.constraint_dict

    new_dict = Dict{Any,ContactConstraint}()
    used_old_keys = Set{Any}()

    # 1. Match New Contacts to Old
    for c in new_contacts
        # Create fresh constraint (stateless)
        con = ContactConstraint(c, friction)

        matched_key = nothing

        # Try to find a match in old constraints
        # Note: Ideally we use spatial hashing lookup, but for exact matching "proximity" logic
        # we might iterate if the exact key drifted. 
        # The Python script iterates `old.items()` to check proximity if exact key logic isn't strict.
        # Here we follow Python's O(N*M) loop inside _proximity_test logic for robustness.

        for (k_old, con_old) in old_dict
            if k_old in used_old_keys
                continue
            end

            if _proximity_test(c, con_old, dist_eps=dist_eps, cos_eps=cos_eps)
                matched_key = k_old

                # Warm Start: Copy lambda and penalty state
                con.lambda_n = con_old.lambda_n
                con.lambda_t1 = con_old.lambda_t1
                con.lambda_t2 = con_old.lambda_t2
                # If you have penalty_k, copy that too

                push!(used_old_keys, k_old)

                # Reset separation counter
                manifold.separation_frames[k_old] = 0
                break
            end
        end

        # Determine Key
        if matched_key !== nothing
            key = matched_key
        else
            # Generate new key
            key = _contact_key(c.point, c.normal)
        end

        new_dict[key] = con

        # Init separation counter if new
        if !haskey(manifold.separation_frames, key)
            manifold.separation_frames[key] = 0
        end
    end

    # 2. Persistence (Keep stale contacts)
    overlap = _aabb_overlap(manifold.bodyA, manifold.bodyB)

    for (k_old, con_old) in old_dict
        if k_old in used_old_keys
            continue
        end

        # Increment age
        sep_age = get(manifold.separation_frames, k_old, 0) + 1

        # If bodies clearly separate, drop immediately (optimization)
        if !overlap
            sep_age = MAX_SEP_FRAMES + 1
        end

        if sep_age <= MAX_SEP_FRAMES
            # Keep the old constraint alive
            new_dict[k_old] = con_old
            manifold.separation_frames[k_old] = sep_age
        else
            # Too old, let it expire (remove from separation tracking)
            delete!(manifold.separation_frames, k_old)
        end
    end

    # 3. Update State
    manifold.constraint_dict = new_dict
    manifold.constraints = collect(values(new_dict))
end

function is_empty(m::Manifold)
    return isempty(m.constraint_dict)
end

end # module